### PR TITLE
Utime

### PR DESCRIPTION
**Pull request by FilippChubukov**

### CODE REVIEW

File: src/compat/posix/stubs/utime.c [hunk 1]
```diff
@@ -9,13 +9,15 @@
...
+#include <fs/kfsop.h>
 
-int utime(const char *path,
-		const struct utimbuf *times) {
-	printk(">>> %s %s %p\n", __func__, path, times);
-	return SET_ERRNO(ENOSYS);
-}
+int utime(const char *path, const struct utimbuf *times) {
 
+	if (-1 == kutime(path, times)) {
```
Comment (anton-bondarev) [inline src/compat/posix/stubs/utime.c]: Why is there a real implementation in stubs folder. You should create a new module for your implementation.

File: src/demo/website/Mybuild [hunk 2]
```diff
@@ -11,8 +11,8 @@ module website {
 		"joke.html",
 		"matmex.gif",
 		"mm_en.gif",
-		"test_temp.html"
-
+		"test_temp.html",
+		"test"
```
Comment (anton-bondarev) [inline src/demo/website/Mybuild]: Why do you use website for adding test file for rootfs? You should use `@InitFS` annotation instead

File: src/compat/posix/include/sys/stat.h [hunk 3]
```diff
@@ -94,10 +94,12 @@ typedef struct stat {
 //	unsigned  st_mtime;   /* time of last modification */
 //	unsigned  st_ctime;   /* time of last status change */
 
+    struct timespec st_crtim; /* time of create */
```
Comment (anton-bondarev) [inline src/compat/posix/include/sys/stat.h]: You should use standard fields for standard structures ` struct stat ` contains  
[https://pubs.opengroup.org/onlinepubs/007904875/basedefs/sys/stat.h.html](https://pubs.opengroup.org/onlinepubs/007904875/basedefs/sys/stat.h.html)
The stat structure shall contain at least the following members:
```
time_t st_mtime;   /* time of last modification */
time_t st_ctime;   /* time of last status change */
```

File: src/compat/posix/include/sys/stat.h [hunk 4]
```diff
@@ -98,6 +98,7 @@ typedef struct stat {
     struct timespec st_mtim;  /* time of last modification */
     struct timespec st_ctim;  /* time of last status change */
 #define st_atime st_atim.tv_sec      /* Backward compatibility */
+#define st_crtime st_crtim.tv_sec
```
Comment (anton-bondarev) [inline src/compat/posix/include/sys/stat.h]: Rm

File: src/include/fs/node.h [hunk 5]
```diff
@@ -56,6 +56,8 @@ typedef struct node {
 
 struct node_info {
 	size_t        size;
+	unsigned int modtime; /* time of modification */
```
Comment (anton-bondarev) [inline src/include/fs/node.h]: it's better mtime and time_t type

File: src/compat/posix/fs/stat.c [hunk 6]
```diff
@@ -6,7 +6,6 @@
  */
 
 #include <sys/stat.h>
-
```
Comment (Deryugin) [inline src/compat/posix/fs/stat.c]: Please, avoid such changes (i.e. you don't actually change anything but just fix formatting), it's not related to your branch and makes it harder to figure out which files were actually changed.

File: src/tests/posix/utime_test.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+TEST_CASE("check changed time") {
+	struct utimbuf prev_time;
+	struct utimbuf new_time;
+
+	struct stat filestat;
+
+	new_time.actime = 999;
+	new_time.modtime = 999;
+
+	creat(TEST_EXIST_FILE, 0666);
```
Comment (Deryugin) [inline src/tests/posix/utime_test.c]: Check for error code? Same for other posix calls

**Comment by Deryugin** [comment]
Don't forget to get latest changes from upstream, it seems that your branch is pretty old.

File: src/fs/syslib/kfsop.c [hunk 8]
```diff
@@ -293,6 +293,20 @@ int klstat(const char *path, struct stat *buf) {
 	return 0;
 }
 
+int kutime(const char *path,const struct utimbuf *times) {
+	struct path node;
+	int res;
+
+	if (0 != (res = fs_perm_lookup(path, NULL, &node))) {
+		errno = -res;
+		return -1;
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/fs/syslib/kfsop.c]: cuz utime return 0 or -1

File: src/fs/syslib/kfile_node.c [hunk 9]
```diff
@@ -64,6 +65,27 @@ int kfile_fill_stat(struct node *node, struct stat *stat_buff) {
...
+	stat_buff->st_mtime = ni->modtime;
+	return 0;
+}
+
+int kfile_change_stat(struct node *node, const struct utimbuf *times) {
+	struct nas *nas;
+	struct node_info *ni;
+	struct timeval now;
+
+	nas = node->nas;
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/fs/syslib/kfile_node.c]: as i understand, its checking in "fs_perm_lookup" function in "kutime".

File: src/compat/posix/utime/utime.c [hunk 10]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @file
+ * @brief
+ *
+ * @author  Anton Kozlov
+ * @date    16.04.2014
+ */
+
+#include <dirent.h>
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/compat/posix/utime/utime.c]: ops, they were before me and I forgot to remove them, my fault

File: src/fs/syslib/kfsop.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (Deryugin) [reply src/fs/syslib/kfsop.c]: But you use `return -errno;` in `utime`

File: src/fs/syslib/kfile_node.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Deryugin) [reply src/fs/syslib/kfile_node.c]: `kfile_change_stat` is not supposed to be called `fs_perm_lookup` (there are simply no reasons why it should be always done this way), so it's a good idea to check this pointer anyway.

File: src/compat/posix/fs/stat.c (hunk 6 above)
Comment (FilippChubukov) [reply src/compat/posix/fs/stat.c]: undo all useless commits

File: src/compat/posix/utime/utime.c [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+#include <errno.h>
+#include <sys/types.h>
+#include <utime.h>
+#include <sys/stat.h>
+#include <fs/kfsop.h>
+
+
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/compat/posix/utime/utime.c]: deleted macro and rmv line

File: src/compat/posix/utime/utime.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/compat/posix/utime/utime.c]: deleted

File: src/fs/syslib/kfsop.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/fs/syslib/kfsop.c]: fixed return in utime(now return -1)

File: src/include/fs/kfsop.h [hunk 12]
```diff
@@ -37,4 +37,8 @@ extern int kfile_fill_stat(struct node *node, struct stat *stat_buff);
 
 extern int kflock(int fd, int operation);
 
+extern int kutime(const char *path,const struct utimbuf *times);
+
+extern int kfile_change_stat(struct node *node,const struct utimbuf *times);
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/include/fs/kfsop.h]: done

File: src/include/fs/node.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/include/fs/node.h]: added

File: src/tests/posix/Mybuild [hunk 13]
```diff
@@ -16,7 +16,11 @@ module all {
 	depends sigchld_test
 	depends waitpid_test
 }
-
+module utime_test {
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/tests/posix/Mybuild]: added

File: src/demo/website/Mybuild [hunk 14]
```diff
@@ -12,7 +12,6 @@ module website {
 		"matmex.gif",
 		"mm_en.gif",
 		"test_temp.html"
-
```
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/demo/website/Mybuild]: undo

File: src/fs/syslib/kfile_node.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (FilippChubukov) [reply src/fs/syslib/kfile_node.c]: added

File: src/tests/posix/utime_test.c (hunk 7 above)
Comment (FilippChubukov) [reply src/tests/posix/utime_test.c]: added

File: src/fs/syslib/kfsop.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (Deryugin) [reply src/fs/syslib/kfsop.c]: Why not just return value here and return it in `utime`? You have to return sane error value anyway according to POSIX

File: src/tests/posix/utime_test.c [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+	test_assert(0 == (fd = creat(TEST_EXIST_FILE, 0666)));
+	test_assert(0 == close(fd));
+
+	test_assert(0 == stat(TEST_EXIST_FILE, &filestat));
+
+	prev_time.actime = filestat.st_ctime;
+	prev_time.modtime = filestat.st_mtime;
+
+	test_assert(utime(TEST_EXIST_FILE, &new_time) == 0);
```
Comment (Deryugin) [inline src/tests/posix/utime_test.c]: Try to use consistent style:

```
assert(0 == f(...))
assert(0 == g(...))
```
or
```
assert(f(...) == 0)
assert(g(...) == 0)
```

but not mix of both.

Comment (FilippChubukov) [reply src/tests/posix/utime_test.c]: fixed

File: src/fs/syslib/kfile_node.c [hunk 16]
```diff
@@ -64,6 +66,32 @@ int kfile_fill_stat(struct node *node, struct stat *stat_buff) {
...
+	return 0;
+}
+
+int kfile_change_stat(struct node *node, const struct utimbuf *times) {
+	struct nas *nas;
+	struct node_info *ni;
+	struct timeval now;
+
+	if (node == NULL) {
+		errno = -ENOENT;
```
Comment (Deryugin) [inline src/fs/syslib/kfile_node.c]: Simply return `-ENOENT`?

File: src/fs/syslib/kfile_node.c [hunk 17]
```diff
@@ -64,6 +66,32 @@ int kfile_fill_stat(struct node *node, struct stat *stat_buff) {
...
+		return -1;
+	}
+
+	nas = node->nas;
+	ni = &nas->fi->ni;
+
+	if (times == NULL) {
+		gettimeofday(&now, NULL);
```
Comment (Deryugin) [inline src/fs/syslib/kfile_node.c]: Check for `gettimeofday()` error

File: src/tests/posix/utime_test.c [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	test_assert(utime(TEST_NOT_EXIST_FILE, &time) == -1);
+}
+
+TEST_CASE("times == NULL") {
+	struct timeval now;
+	struct stat filestat;
+
+	int fd;
+
+	test_assert((fd = creat(TEST_EXIST_FILE, 0666)) == 0);
```
Comment (Deryugin) [inline src/tests/posix/utime_test.c]: Why `fd` should be zero?

File: src/fs/syslib/kfile_node.c (hunk 16 above)
Comment (FilippChubukov) [reply src/fs/syslib/kfile_node.c]: done

File: src/fs/syslib/kfile_node.c (hunk 17 above)
Comment (FilippChubukov) [reply src/fs/syslib/kfile_node.c]: done

File: src/tests/posix/utime_test.c (hunk 18 above)
Comment (FilippChubukov) [reply src/tests/posix/utime_test.c]: ye, reworked this place, fd mb >= 0

**Comment by Deryugin** [comment]
Please, add stub version of `kutime` to `embox.fs.syslib.file_system_none` module.

File: src/compat/posix/utime/Mybuild [hunk 19]
```diff
@@ -0,0 +1,5 @@
+package embox.compat.posix
+
+static module utime {
+	source "utime.c"
```
Comment (Deryugin) [inline src/compat/posix/utime/Mybuild]: Please, add module dependency (for `kutime` function).

File: src/tests/posix/Mybuild [hunk 20]
```diff
@@ -17,6 +17,11 @@ module all {
 	depends waitpid_test
 }
 
+module utime_test {
```
Comment (Deryugin) [inline src/tests/posix/Mybuild]: Please, add this test to some template or to the `embox.test.posix.all` package to make sure it's used somewhere.

**Comment by Deryugin** [comment]
Tests fail for x86/qemu template

```
	failure at src/tests/posix/utime_test.c : 30, in function __test_case_at_line_20
		test_assert((fd = creat(TEST_EXIST_FILE, 0666)) >= 0)
	   case at src/tests/posix/utime_test.c : 20
		"check changed time"
	.
	failure at src/tests/posix/utime_test.c : 65, in function __test_case_at_line_59
		test_assert((fd = creat(TEST_EXIST_FILE, 0666)) >= 0)
	   case at src/tests/posix/utime_test.c : 59
		"times == NULL"
	
	testing utime_test (posix/utime_test) failed
		2/3 failures
```

File: src/compat/posix/utime/Mybuild (hunk 19 above)
Comment (FilippChubukov) [reply src/compat/posix/utime/Mybuild]: added kutime in stub version, added depends of kutime in utime Mybuild

File: src/tests/posix/Mybuild (hunk 20 above)
Comment (FilippChubukov) [reply src/tests/posix/Mybuild]: added in tempate of x86/qemu

Comment (FilippChubukov) [reply src/tests/posix/Mybuild]: set runlevel(2) for utime_test and its working

Comment (FilippChubukov) [reply src/tests/posix/Mybuild]: last commit failed in travis, but it looks like an error in something not related to my commit, please see how will the time
"Error: Gold frame ./scripts/continuous/qt/gold_moveblocks_frames/3.jpg was not found in ./scripts/continuous/qt/captured.mpeg
- - - - - - - - - - - - - - -
ERROR: Test Qt moveblocks over vnc FAILED
- - - - - - - - - - - - - - -
error:
"

**Pull request merged**